File: backend-go/internal/mail/mail.go
// Package mail sends the share/mention notification emails required by
// the product spec, over plain SMTP.
package mail

import (
	"fmt"
	"log/slog"
	"net/smtp"
)

type Mailer struct {
	host, port, username, password, from string
}

func New(host, port, username, password, from string) *Mailer {
	return &Mailer{host: host, port: port, username: username, password: password, from: from}
}

func (m *Mailer) send(to, subject, body string) error {
	addr := fmt.Sprintf("%s:%s", m.host, m.port)
	msg := fmt.Sprintf("From: %s\r\nTo: %s\r\nSubject: %s\r\nContent-Type: text/plain; charset=UTF-8\r\n\r\n%s\r\n",
		m.from, to, subject, body)

	var auth smtp.Auth
	if m.username != "" {
		auth = smtp.PlainAuth("", m.username, m.password, m.host)
	}
	if err := smtp.SendMail(addr, auth, m.from, []string{to}, []byte(msg)); err != nil {
		return fmt.Errorf("send mail to %s: %w", to, err)
	}
	return nil
}

// sendBestEffort logs failures instead of returning them: a notification
// email failing to send must never fail the API request that triggered it
// (e.g. saving a note the user just wrote).
func (m *Mailer) sendBestEffort(to, subject, body string) {
	if err := m.send(to, subject, body); err != nil {
		slog.Error("failed to send notification email", "to", to, "error", err)
	}
}

func (m *Mailer) NotifyNoteShared(to, actorName, noteTitle, noteURL string, editable bool) {
	access := "read-only"
	if editable {
		access = "editable"
	}
	subject := fmt.Sprintf("%s shared a note with you: %s", actorName, noteTitle)
	body := fmt.Sprintf("%s shared the note %q with you (%s access).\n\nOpen it here: %s\n", actorName, noteTitle, access, noteURL)
	m.sendBestEffort(to, subject, body)
}

func (m *Mailer) NotifyMentioned(to, actorName, noteTitle, noteURL string) {
	subject := fmt.Sprintf("%s mentioned you in a note: %s", actorName, noteTitle)
	body := fmt.Sprintf("%s mentioned you in the note %q.\n\nOpen it here: %s\n", actorName, noteTitle, noteURL)
	m.sendBestEffort(to, subject, body)
}

File: backend-python/app/mail.py
"""Sends the share/mention notification emails required by the product
spec, over plain SMTP (stdlib smtplib — no extra dependency needed).
"""

from __future__ import annotations

import logging
import smtplib
from email.message import EmailMessage

logger = logging.getLogger("app.mail")


class Mailer:
    def __init__(self, host: str, port: int, username: str, password: str, from_addr: str) -> None:
        self._host = host
        self._port = port
        self._username = username
        self._password = password
        self._from = from_addr

    def _send(self, to: str, subject: str, body: str) -> None:
        msg = EmailMessage()
        msg["From"] = self._from
        msg["To"] = to
        msg["Subject"] = subject
        msg.set_content(body)

        with smtplib.SMTP(self._host, self._port, timeout=10) as smtp:
            if self._username:
                smtp.login(self._username, self._password)
            smtp.send_message(msg)

    def _send_best_effort(self, to: str, subject: str, body: str) -> None:
        """A notification email failing to send must never fail the API
        request that triggered it (e.g. saving a note the user just wrote).
        """
        try:
            self._send(to, subject, body)
        except OSError:
            logger.exception("failed to send notification email to %s", to)

    def notify_note_shared(
        self, to: str, actor_name: str, note_title: str, note_url: str, editable: bool
    ) -> None:
        access = "editable" if editable else "read-only"
        subject = f"{actor_name} shared a note with you: {note_title}"
        body = (
            f'{actor_name} shared the note "{note_title}" with you ({access} access).\n\n'
            f"Open it here: {note_url}\n"
        )
        self._send_best_effort(to, subject, body)

    def notify_mentioned(self, to: str, actor_name: str, note_title: str, note_url: str) -> None:
        subject = f"{actor_name} mentioned you in a note: {note_title}"
        body = f'{actor_name} mentioned you in the note "{note_title}".\n\nOpen it here: {note_url}\n'
        self._send_best_effort(to, subject, body)
